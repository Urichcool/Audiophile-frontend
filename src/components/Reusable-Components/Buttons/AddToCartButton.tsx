import { FC } from "react";

interface IAddToCartButtonProps{
  testId?:string
}

const AddToCartButton: FC<IAddToCartButtonProps> = ({testId}) => {
  return (
    <button className="add-to-cart-button" data-testid={testId} type="submit">
      Add to cart
    </button>
  );
};

export default AddToCartButton;
