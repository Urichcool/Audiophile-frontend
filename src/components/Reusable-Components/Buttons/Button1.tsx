import { FC } from "react";
import { NavigateFunction, useNavigate } from "react-router-dom";

interface IButton1Props {
  id?: string;
  testId: string;
}

const Button1: FC<IButton1Props> = ({ id, testId }) => {
  const navigate: NavigateFunction = useNavigate();

  const buttonClickHandler = (e: React.MouseEvent<HTMLButtonElement>): void => {
    navigate(`/${id}`);
  };

  return (
    <button
      className="button1"
      data-testid={testId}
      onClick={buttonClickHandler}
    >
      See product
    </button>
  );
};

export default Button1;
