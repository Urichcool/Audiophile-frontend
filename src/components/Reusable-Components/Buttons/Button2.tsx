import { FC } from 'react';
import { NavigateFunction, useNavigate } from "react-router-dom";

interface IButton2Props {
  id?: string;
  testId?:string
}

const Button2: FC<IButton2Props> = ({id, testId}) => {
  const navigate: NavigateFunction = useNavigate();

 const buttonClickHandler = (e: React.MouseEvent<HTMLButtonElement>): void => {
   navigate(`/${id}`);
 };

    return (
      <button className='button2' data-testid={testId} onClick={buttonClickHandler}>see product</button>
    );
}

export default Button2;