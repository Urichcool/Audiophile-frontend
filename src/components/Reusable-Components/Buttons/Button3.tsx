import { FC } from "react";
import { useNavigate } from "react-router-dom";
import ArrowRightIcon from "../../../images/icons/ArrowRightIcon";
import { AppDispatch } from "../../../redux/store";
import {
  useAppDispatch,
  useAppSelector,
} from "../../../redux/reduxHooks/reduxHooks";
import {
  selectIsMenuOpen,
  switchMenu,
} from "../../../redux/slices/mobileMenuSlice";

interface IButton3Props {
  to: "headphones" | "speakers" | "earphones";
  isMobile: boolean | undefined;
}

const Button3: FC<IButton3Props> = ({ to, isMobile }) => {
  const dispatch: AppDispatch = useAppDispatch();
  const isMenuOpen: boolean = useAppSelector(selectIsMenuOpen);
  const navigate = useNavigate();

  return (
    <button
      className="button3"
      onClick={():void => {
        navigate(to);
        if (isMobile) {
          dispatch(switchMenu(!isMenuOpen));
        }
      }}
    >
      <p>Shop</p>
      <ArrowRightIcon />
    </button>
  );
};

export default Button3;
