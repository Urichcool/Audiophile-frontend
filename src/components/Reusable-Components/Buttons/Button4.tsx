import { FC } from 'react';
import { NavigateFunction, useNavigate } from "react-router-dom";

interface IButton4Props {
  id?: string ;
}

const Button4: FC<IButton4Props> = ({ id }) => {
  const navigate: NavigateFunction = useNavigate();
  
  const buttonClickHandler = (e: React.MouseEvent<HTMLButtonElement>): void => {
    navigate(`/${id}`);
  };
  
    return ( 
     <button className='button4' onClick={buttonClickHandler}>See Product</button>
    );
}

export default Button4;