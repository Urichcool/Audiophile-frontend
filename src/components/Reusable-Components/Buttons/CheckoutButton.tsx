import React, { FC } from "react";
import { NavigateFunction, useNavigate } from "react-router-dom";
import {
  useAppDispatch,
  useAppSelector,
} from "../../../redux/reduxHooks/reduxHooks";
import {
  selectCartProducts,
  selectIsCartModalOpen,
} from "../../../redux/slices/cart/selectors";
import { AppDispatch } from "../../../redux/store";
import { switchCartModal } from "../../../redux/slices/cart/cartSlice";
import { Oval } from "react-loader-spinner";
import { useCheckGoodsCartStockMutation } from "../../../redux/services/goods";
import { FetchBaseQueryError } from "@reduxjs/toolkit/query";
import { SerializedError } from "@reduxjs/toolkit";

interface ICheckoutButton {
  testId?: string;
  isFetching: boolean;
}

const CheckoutButton: FC<ICheckoutButton> = ({ testId, isFetching }) => {
  const navigate: NavigateFunction = useNavigate();
  const isCartModalOpen: boolean = useAppSelector(selectIsCartModalOpen);
  const dispatch: AppDispatch = useAppDispatch();
  const [checkCartStock, { isLoading }] =
    useCheckGoodsCartStockMutation();
  const cart: {
    id: string;
    name: string;
    quantity: number;
    price: number;
    totalPrice: number;
    picture: string;
    category: string;
  }[] = useAppSelector(selectCartProducts);

  const buttonClickHandler = async (e: React.MouseEvent<HTMLButtonElement>) => {
    const result:
      | {
          data: {
            isEnoughCartStock: boolean;
          };
        }
      | {
          error: FetchBaseQueryError | SerializedError;
        } = await checkCartStock(
      cart.map(({ id, quantity }) => {
        return { id: id, quantity: quantity };
      })
    );
    if (Object.values(result)[0].isEnoughCartStock) {
      navigate("checkout");
      dispatch(switchCartModal(!isCartModalOpen));
    }
  };
  return (
    <>
      <button
        onClick={buttonClickHandler}
        className="checkout-button"
        data-testid={testId}
        disabled={isFetching}
      >
        {isFetching || isLoading ? (
          <div className="checkout-button-loader">
            <p>Checkout</p>
            <Oval
              height={16}
              width={16}
              color="#fff"
              wrapperStyle={{}}
              wrapperClass=""
              visible={true}
              ariaLabel="oval-loading"
              secondaryColor="#fff"
              strokeWidth={2}
              strokeWidthSecondary={2}
            />
          </div>
        ) : (
          <p>Checkout</p>
        )}
      </button>
    </>
  );
};

export default CheckoutButton;
