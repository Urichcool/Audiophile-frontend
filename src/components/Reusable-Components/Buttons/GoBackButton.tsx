import { FC } from "react";
import { NavigateFunction, useNavigate } from "react-router-dom";

interface IGoBackButtonProps {
  testId?: string;
}

const GoBackButton: FC<IGoBackButtonProps> = ({ testId }) => {
  const navigate: NavigateFunction = useNavigate();

  const buttonClickHandler = (e: React.MouseEvent<HTMLButtonElement>): void => {
    navigate(-1);
  };

  return (
    <button
      data-testid={testId}
      className="Body-manrope-medium go-back-button"
      onClick={buttonClickHandler}
    >
      Go back
    </button>
  );
};

export default GoBackButton;
