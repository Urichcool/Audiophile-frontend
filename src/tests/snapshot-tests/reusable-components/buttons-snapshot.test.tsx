import AddToCartButton from "../../../components/Reusable-Components/Buttons/AddToCartButton";
import Button1 from "../../../components/Reusable-Components/Buttons/Button1";
import Button2 from "../../../components/Reusable-Components/Buttons/Button2";
import Button3 from "../../../components/Reusable-Components/Buttons/Button3";
import Button4 from "../../../components/Reusable-Components/Buttons/Button4";
import CheckoutButton from "../../../components/Reusable-Components/Buttons/CheckoutButton";
import GoBackButton from "../../../components/Reusable-Components/Buttons/GoBackButton";
import { renderWithReduxAndRouter } from "../../helpers/RenderWithReduxAndRouter";
import { screen } from "@testing-library/react";

describe("buttons-render-tests", () => {
  test("add-to-cart-button-should-renders-correctly", () => {
    renderWithReduxAndRouter(
      <AddToCartButton testId="add-to-cart-test-button" />
    );
    const addToCartButton: HTMLButtonElement = screen.getByTestId(
      "add-to-cart-test-button"
    );
    expect(addToCartButton).toMatchSnapshot();
  });
  test("button-1-should-match-snapshot", () => {
    renderWithReduxAndRouter(<Button1 testId="test-button-1" />);
    const button1: HTMLButtonElement = screen.getByTestId("test-button-1");
    expect(button1).toMatchSnapshot();
  });
  test("button-2-should-match-snapshot", () => {
    renderWithReduxAndRouter(<Button2 testId="test-button-2" />);
    const button2: HTMLButtonElement = screen.getByTestId("test-button-2");
    expect(button2).toMatchSnapshot();
  });
  test("button-3-should-match-snapshot", () => {
    renderWithReduxAndRouter(
      <Button3 testId="test-button-3" to="/headphones" />
    );
  });
  test("button-4-should-match-snapshot", () => {
    renderWithReduxAndRouter(<Button4 testId="test-button-4" />);
    const button4: HTMLButtonElement = screen.getByTestId("test-button-4");
    expect(button4).toMatchSnapshot();
  });
  test("checkout-button-should-match-snapshot", () => {
    renderWithReduxAndRouter(<CheckoutButton testId="checkout-test-button" />);
    const checkoutButton: HTMLButtonElement = screen.getByTestId(
      "checkout-test-button"
    );
    expect(checkoutButton).toMatchSnapshot();
  });
  test("go-back-button-should-match-snapshot", () => {
    renderWithReduxAndRouter(<GoBackButton testId="go-back-test-button" />);
    const goBackButton: HTMLButtonElement = screen.getByTestId(
      "go-back-test-button"
    );
    expect(goBackButton).toMatchSnapshot();
  });
});
