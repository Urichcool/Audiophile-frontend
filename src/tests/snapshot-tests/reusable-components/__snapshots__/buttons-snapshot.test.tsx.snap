// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buttons-render-tests add-to-cart-button-should-renders-correctly 1`] = `
<button
  class="add-to-cart-button"
  data-testid="add-to-cart-test-button"
  type="submit"
>
  Add to cart
</button>
`;

exports[`buttons-render-tests button-1-should-match-snapshot 1`] = `
<button
  class="button1"
  data-testid="test-button-1"
>
  See product
</button>
`;

exports[`buttons-render-tests button-2-should-match-snapshot 1`] = `
<button
  class="button2"
  data-testid="test-button-2"
>
  see product
</button>
`;

exports[`buttons-render-tests button-4-should-match-snapshot 1`] = `
<button
  class="button4"
  data-testid="test-button-4"
>
  See Product
</button>
`;

exports[`buttons-render-tests checkout-button-should-match-snapshot 1`] = `
<button
  class="checkout-button"
  data-testid="checkout-test-button"
>
  Checkout
</button>
`;

exports[`buttons-render-tests go-back-button-should-match-snapshot 1`] = `
<button
  class="Body-manrope-medium go-back-button"
  data-testid="go-back-test-button"
>
  Go back
</button>
`;
